### ISSUE TITLE
Failure to open streams does not report a very helpful error

### ISSUE THREAD

**Issue by danpat** [Project-OSRM/osrm-backend#3047]
https://github.com/Project-OSRM/osrm-backend/blob/77931ffe2405b579c0e9c8cd8772957aef75a114/src/contractor/contractor.cpp#L712

The root cause of this error is not displayed.  It could be a permissions problem, it could be a missing file, or some other error.

We can cause `std::ofstream` to throw an exception on error by using the default constructor, setting `stream.exceptions(std::ifstream::failbit | std::ifstream::badbit);`, then using `stream.open(filename)`.  The thrown exception's `what()` will then contain a more detailed error message (like `Permission denied`, or `File not found`).

In theory, the global `errno` should be thread-local and contain the most recent syscall error (and be convertible to a string with `std::strerror()`, but there are various reports floating around that that's less reliable than using the exception approach.

**Comment by danpat** [comment]
Simple example:

Instead of:

```
std::ofstream foo(filename);
if (!foo) { throw error...}
```

we should do:
```
std::ofstream foo;
foo.exceptions(std::ofstream::failbit | std::ofstream::badbit);
try {
  foo.open(filename);
} catch (const someexception &e) {
  std::err << "Failed to open file " << filename << " because " << e.what() << std::endl;
  throw error;
}
```

**Comment by TheMarex** [comment]
Can we write our own wrapper of `std::stream` for this? This seems like a lot of code duplication.

**Comment by daniel-j-h** [comment]
There is no reason to use streams _at all_ for non-formatted input / output.
All we're ever doing is reading / writing _bytes_.

Check rtd for some nice abstractions for `File` / `Stream`.

**Comment by danpat** [comment]
Upon doing some investigation for #3165 , it turns out that `e.what()` doesn't report `strerr` at all on many platforms, making the exception-catching approach not quite as attractive.

**Comment by danpat** [comment]
Fixed with #3321
